From ff1fed69f3ce7773bb84599dbe4ec45012248d4c Mon Sep 17 00:00:00 2001
From: Madhavan Ganesan <quic_madhgane@quicinc.com>
Date: Tue, 3 Dec 2024 11:12:52 +0530
Subject: [PATCH] UPSTREAM:CP:wifi:ath12k: Add pdev_id param to FTM WMI to
 handle split PHY

In FTM(Factory Test Mode) mode, the SOC level pdev_id received instead of
the radio level pdev_id from the FTM segment header. The split PHY target
has PHYA0 and PHYA1 in a SOC. Therefore, we need to obtain a unique pdev_id
for the split PHY to send and receive data properly based on the pdev_id.

To fix this issue, add the pdev_id param in FTM WMI command based on the
service bit to handle the split PHY use cases.

Signed-off-by: Madhavan Ganesan <quic_madhgane@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/testmode.c | 62 ++++++++++++++++++----
 drivers/net/wireless/ath/ath12k/testmode.h |  6 +--
 drivers/net/wireless/ath/ath12k/wmi.c      | 10 +---
 drivers/net/wireless/ath/ath12k/wmi.h      | 13 +++++
 4 files changed, 67 insertions(+), 24 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/testmode.c
+++ b/drivers/net/wireless/ath/ath12k/testmode.c
@@ -155,24 +155,43 @@ out:
 * Data of various events received from fw is aggregated and
 * sent to application layer
 */
-int ath12k_tm_process_event(struct ath12k_base *ab, u32 cmd_id,
-			    const struct wmi_ftm_event_msg *ftm_msg,
-                            u16 length)
+int ath12k_tm_process_event(struct ath12k_base *ab, u32 cmd_id, const void **tb,
+			    u16 length)
 {
+	const struct wmi_pdev_utf_event_param *param = NULL;
+	const struct wmi_ftm_event_msg *ftm_msg;
+	u8 total_segments, current_seq;
 	struct sk_buff *nl_skb;
-	int ret = 0;
 	struct ath12k *ar;
 	u8 const *buf_pos;
-	u16 datalen;
-	u8 total_segments, current_seq;
 	u32 data_pos;
+	int ret = 0;
+	u16 datalen;
 	u32 pdev_id;
 
+	ftm_msg = tb[WMI_TAG_ARRAY_BYTE];
+	if (!ftm_msg) {
+		ath12k_warn(ab, "failed to fetch ftm msg\n");
+		return -EINVAL;
+	}
+
 	ath12k_dbg(ab, ATH12K_DBG_TESTMODE,
 		   "testmode event wmi cmd_id %d ftm event msg %pK datalen %d\n",
 		   cmd_id, ftm_msg, length);
 	ath12k_dbg_dump(ab, ATH12K_DBG_TESTMODE, NULL, "", ftm_msg, length);
-	pdev_id = DP_HW2SW_MACID(ftm_msg->seg_hdr.pdev_id);
+
+	if (test_bit(WMI_TLV_SERVICE_PDEV_PARAM_IN_UTF_WMI, ab->wmi_ab.svc_map)) {
+		param = tb[WMI_TAG_PDEV_UTF_EVENT_FIXED_PARAM];
+		if (!param) {
+			ath12k_warn(ab, "failed to fetch utf msg\n");
+			return -EINVAL;
+		}
+		length -= (sizeof(*param) + TLV_HDR_SIZE);
+		pdev_id = le32_to_cpu(DP_HW2SW_MACID(param->pdev_id));
+		ath12k_dbg_dump(ab, ATH12K_DBG_TESTMODE, NULL, "", param, sizeof(*param));
+	} else {
+		pdev_id = DP_HW2SW_MACID(ftm_msg->seg_hdr.pdev_id);
+	}
 
 	if (pdev_id >= ab->num_radios) {
 		ath12k_warn(ab, "testmode event not handled due to invalid pdev id\n");
@@ -401,11 +420,13 @@ out:
 
 static int ath12k_tm_cmd_process_ftm(struct ath12k *ar, struct nlattr *tb[])
 {
+	struct wmi_pdev_utf_cmd_fixed_param *utf_cmd;
 	struct ath12k_wmi_pdev *wmi = ar->wmi;
-	struct sk_buff *skb;
 	u32 cmd_id, buf_len, hdr_info;
-	int ret;
+	struct sk_buff *skb;
+	uint16_t len;
 	void *buf;
+	int ret;
 
 	/* if buf_len is 0 no data is sent, return error */
 	u8 segnumber = 0, seginfo;
@@ -443,9 +464,12 @@ static int ath12k_tm_cmd_process_ftm(str
 		else
 			chunk_len = buf_len;
 
-		skb = ath12k_wmi_alloc_skb(wmi->wmi_ab, (chunk_len +
-					sizeof(struct wmi_ftm_cmd)));
+		if (test_bit(WMI_TLV_SERVICE_PDEV_PARAM_IN_UTF_WMI, ar->ab->wmi_ab.svc_map))
+			len = chunk_len + sizeof(*ftm_cmd) + sizeof(*utf_cmd);
+		else
+			len = chunk_len + sizeof(*ftm_cmd);
 
+		skb = ath12k_wmi_alloc_skb(wmi->wmi_ab, len);
 		if (!skb) {
 			ret = -ENOMEM;
 			goto out;
@@ -463,6 +487,13 @@ static int ath12k_tm_cmd_process_ftm(str
 		ftm_cmd->seg_hdr.segmentinfo = seginfo;
 		segnumber++;
 		memcpy(&ftm_cmd->data, bufpos, chunk_len);
+		if (test_bit(WMI_TLV_SERVICE_PDEV_PARAM_IN_UTF_WMI, ar->ab->wmi_ab.svc_map)) {
+			utf_cmd = (struct wmi_pdev_utf_cmd_fixed_param *)(skb->data + chunk_len + sizeof(*ftm_cmd));
+			utf_cmd->tlv_header = FIELD_PREP(WMI_TLV_TAG, WMI_TAG_PDEV_UTF_CMD_FIXED_PARAM) |
+				FIELD_PREP(WMI_TLV_LEN, sizeof(*utf_cmd) - TLV_HDR_SIZE);
+			utf_cmd->pdev_id = cpu_to_le32(ar->pdev->pdev_id);
+		}
+
 		ret = ath12k_wmi_cmd_send(wmi, skb, cmd_id);
 		if (ret) {
 			ath12k_warn(ar->ab, "ftm wmi command fail: %d\n", ret);
--- a/drivers/net/wireless/ath/ath12k/testmode.h
+++ b/drivers/net/wireless/ath/ath12k/testmode.h
@@ -10,8 +10,7 @@
 #ifdef CPTCFG_NL80211_TESTMODE
 
 void ath12k_tm_wmi_event_unsegmented(struct ath12k_base *ab, u32 cmd_id, struct sk_buff *skb);
-int ath12k_tm_process_event(struct ath12k_base *ab, u32 cmd_id,
-			    const struct wmi_ftm_event_msg *ftm_msg,
+int ath12k_tm_process_event(struct ath12k_base *ab, u32 cmd_id, const void **tb,
 			    u16 length);
 int ath12k_tm_cmd(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 		  u8 link_id, void *data, int len);
@@ -25,8 +24,7 @@ static inline void ath12k_tm_wmi_event_u
 }
 
 static inline int ath12k_tm_process_event(struct ath12k_base *ab, u32 cmd_id,
-					  const struct wmi_ftm_event_msg *msg,
-					  u16 length)
+					  const void **tb, u16 length)
 {
 	return 0;
 }
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -14484,7 +14484,6 @@ static void ath12k_wmi_tm_event_segmente
 					struct sk_buff *skb)
 {
 	const void **tb;
-	const struct wmi_ftm_event_msg *ev;
 	u16 length;
 	int ret;
 
@@ -14496,15 +14495,8 @@ static void ath12k_wmi_tm_event_segmente
 		return;
 	}
 
-	ev = tb[WMI_TAG_ARRAY_BYTE];
-	if (!ev) {
-		ath12k_warn(ab, "failed to fetch ftm msg\n");
-		kfree(tb);
-		return;
-	}
-
 	length = skb->len - TLV_HDR_SIZE;
-	ret = ath12k_tm_process_event(ab, cmd_id, ev, length);
+	ret = ath12k_tm_process_event(ab, cmd_id, tb, length);
 	if (ret)
 		ath12k_warn(ab, "Failed to process ftm event\n");
 	kfree(tb);
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -2277,6 +2277,8 @@ enum wmi_tlv_tag {
 	WMI_TAG_MLO_LINK_REMOVAL_CMD_FIXED_PARAM = 0x464,
 	WMI_TAG_PDEV_WSI_STATS_INFO_CMD = 0x4b1,
 	WMI_TAG_PDEV_DFS_RADAR_FLAGS = 0x4b4,
+	WMI_TAG_PDEV_UTF_CMD_FIXED_PARAM = 0x4be,
+	WMI_TAG_PDEV_UTF_EVENT_FIXED_PARAM,
 	WMI_TAG_TWT_VDEV_CONFIG_CMD = 0x4de,
 	WMI_TAG_PDEV_POWER_BOOST_EVENT_FIXED_PARAM = 0x4E8,
 	WMI_TAG_PDEV_POWER_BOOST_CMD_FIXED_PARAM,
@@ -2526,6 +2528,7 @@ enum wmi_tlv_service {
 	WMI_TLV_SERVICE_BRIDGE_VDEV_SUPPORT = 384,
 	WMI_TLV_SERVICE_PDEV_WSI_STATS_INFO_SUPPORT = 388,
 	WMI_TLV_SERVICE_RADAR_FLAGS_SUPPORT = 390,
+	WMI_TLV_SERVICE_PDEV_PARAM_IN_UTF_WMI = 394,
 	WMI_TLV_SERVICE_SW_PROG_DFS_SUPPORT = 395,
 	WMI_TLV_SERVICE_THERM_THROT_POUT_REDUCTION = 410,
 	WMI_SERVICE_WDS_NULL_FRAME_SUPPORT = 421,
@@ -4595,6 +4598,16 @@ struct wmi_ftm_event_msg {
 	struct wmi_ftm_seg_hdr seg_hdr;
 	u8 data[];
 } __packed;
+
+struct wmi_pdev_utf_cmd_fixed_param {
+	__le32 tlv_header;
+	__le32 pdev_id;
+};
+
+struct wmi_pdev_utf_event_param {
+	__le32 pdev_id;
+};
+
 #define WMI_BEACON_TX_BUFFER_SIZE	512
 
 #define WMI_BEACON_EMA_PARAM_PERIODICITY_SHIFT         0
